### DISCUSSION TITLE
How to fix high dpi in WPF?

### DISCUSSION THREAD

**Discussion by VahidEra** [dotnet/wpf#7455] [Q&A]
I have a problem with high dpi in WPF project, it is like this in design:

https://1drv.ms/u/s!AqwLh1F7nIIAgxYdEQ8u6vCkMUn8?e=4lJeYM

But when I run the program it is like this:

https://1drv.ms/u/s!AqwLh1F7nIIAgxeonoffPEG7KkOg?e=eSyePg

Considering that I also changed the manifest file, the problem is still not solved:
```
<?xml version="1.0" encoding="utf-8"?>
<assembly manifestVersion="1.0" xmlns="urn:schemas-microsoft-com:asm.v1">
  <assemblyIdentity version="1.0.0.0" name="MyApplication.app"/>
 <compatibility xmlns="urn:schemas-microsoft-com:compatibility.v1">
    <application>
      <!-- A list of the Windows versions that this application has been tested on
           and is designed to work with. Uncomment the appropriate elements
           and Windows will automatically select the most compatible environment. -->

      <!-- Windows Vista -->
      <supportedOS Id="{e2011457-1546-43c5-a5fe-008deee3d3f0}" />

      <!-- Windows 7 -->
      <supportedOS Id="{35138b9a-5d96-4fbd-8e2d-a2440225f93a}" />

      <!-- Windows 8 -->
      <supportedOS Id="{4a2f28e3-53b9-4441-ba9c-d69d4a4a6e38}" />

      <!-- Windows 8.1 -->
      <supportedOS Id="{1f676c76-80e1-4239-95bb-83d0f6d0da78}" />

      <!-- Windows 10 -->
      <supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}" />

    </application>
  </compatibility>
<application xmlns="urn:schemas-microsoft-com:asm.v3">
    <windowsSettings>
      <dpiAware xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">true</dpiAware>
      <longPathAware xmlns="http://schemas.microsoft.com/SMI/2016/WindowsSettings">true</longPathAware>
    </windowsSettings>
  </application>
</assembly>
```
>**Guys please help me in this matter thank you very much for your help**

**Comment by VahidEra** [comment]
@miloush Dear miloush, please take a look at this question. Thank you very much for your kindness and help.

**Reply by miloush** [reply]
You have to understand that WPF is very different from Winforms and one of the larger differences is in the positioning of controls. You might find https://social.technet.microsoft.com/wiki/contents/articles/12898.moving-from-winforms-to-wpf.aspx useful.

In short, instead of using absolute dimensions and coordinates, you should let the framework position the elements when they fit. You use panels to do that, check the list here: https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/layout?view=netframeworkdesktop-4.8#panel-elements-and-custom-layout-behaviors

In your case, you can do something like
```xml
<Grid>
    <Grid.RowDefinitions>
        <RowDefinition />
        <RowDefinition Height="Auto" />
    </Grid.RowDefinitions>
    ...
    <StackPanel Grid.Row="1" Orientation="Horizontal" HorizontalAlignment="Center">
        <Button Content="Button" Padding="5" MinWidth="80" />
        <Button Content="Button" Padding="5" MinWidth="80" Margin="100,0,0,0" /> <!-- Or use Grid here too and have the gap relatively sized to the buttons -->
    </StackPanel>
</Grid>
```

Alternatively, you can use DockPanel:
```xml
<DockPanel>
    <StackPanel DockPanel.Dock="Bottom" Orientation="Horizontal" HorizontalAlignment="Center">
        <Button Content="Button" Padding="5" MinWidth="80" />
        <Button Content="Button" Padding="5" MinWidth="80" Margin="100,0,0,0" />
    </StackPanel>
    <Grid>
        ...
    </Grid>
</DockPanel>
```

In both ways, the button row will take as much space as needed from the bottom of the window and rest will get the remaining space. I don't think you are really having a DPI issue here.

**Comment by VahidEra** [comment]
@miloush Dear miloush,
Now I understand.  You are exactly right, the main problem is the size of the grid and stackpanel And there is no problem with high dpi settings.
This is exactly what I wanted
Thank you very much for your kindness and help, dear friend.
